ee/server/service/calendar.go

```diff
@@ -129,5 +191,151 @@ func (svc *Service) CalendarWebhook(ctx context.Context, eventUUID string, chann
+		// the queue (actually a set) to indicate that we need to re-process the event.
+		err = svc.distributedLock.AddToSet(ctx, calendar.QueueKey, eventUUID)
+		if err != nil {
+			return "", false, ctxerr.Wrap(ctx, err, "add calendar event to queue")
+		}
+
+		if reserved {
+			// We flag the lock as reserved.
+			return "", reserved, nil
+		}
+
+		// Try to acquire the lock again in case it was released while we were adding the event to the queue.
+		result, err = svc.distributedLock.AcquireLock(ctx, calendar.LockKeyPrefix+eventUUID, lockValue, 0)
+		if err != nil {
+			return "", false, ctxerr.Wrap(ctx, err, "acquire calendar lock again")
+		}
+
+		if result == "" {
+			// We could not acquire the lock, so we are done here.
+			return "", reserved, nil
+		}
+	}
+	return lockValue, false, nil
+}
+
+func (svc *Service) processCalendarAsync(ctx context.Context, eventIDs []string) {
+	defer func() {
+		asyncMutex.Lock()
+		asyncCalendarProcessing = false
+		asyncMutex.Unlock()
+	}()
+	for {
+		if len(eventIDs) == 0 {
+			return
+		}
+		for _, eventUUID := range eventIDs {
+			if ok := svc.processCalendarEventAsync(ctx, eventUUID); !ok {
+				return
+			}
+		}
+
+		// Now we check whether there are any more events in the queue.
+		var err error
+		eventIDs, err = svc.distributedLock.GetSet(ctx, calendar.QueueKey)
+		if err != nil {
+			level.Error(svc.logger).Log("msg", "Failed to get calendar event queue", "err", err)
+			return
+		}
+	}
+}
+
+func (svc *Service) processCalendarEventAsync(ctx context.Context, eventUUID string) bool {
+	lockValue, _, err := svc.getCalendarLock(ctx, eventUUID, false)
+	if err != nil {
+		level.Error(svc.logger).Log("msg", "Failed to get calendar lock", "err", err)
+		return false
+	}
+	if lockValue == "" {
+		// We did not get a lock, so there is nothing to do here
+		return true
+	}
+	defer svc.releaseCalendarLock(ctx, eventUUID, lockValue)
+
+	// Remove event from the queue so that we don't process this event again.
+	// Note: This item can be added back to the queue while we are processing it.
+	err = svc.distributedLock.RemoveFromSet(ctx, calendar.QueueKey, eventUUID)
+	if err != nil {
+		level.Error(svc.logger).Log("msg", "Failed to remove calendar event from queue", "err", err)
+		return false
+	}
+
+	appConfig, err := svc.ds.AppConfig(ctx)
+	if err != nil {
+		level.Error(svc.logger).Log("msg", "Failed to load app config", "err", err)
+		return false
+	}
+
+	if len(appConfig.Integrations.GoogleCalendar) == 0 {
+		// Google Calendar integration is not configured
+		return true
+	}
+	googleCalendarIntegrationConfig := appConfig.Integrations.GoogleCalendar[0]
+
+	eventDetails, err := svc.ds.GetCalendarEventDetailsByUUID(ctx, eventUUID)
+	if err != nil {
+		if fleet.IsNotFound(err) {
+			// We found this event when the callback initially came in. So the event may have been removed from DB since then.
+			return true
```

Comment:
This happens frequently because we create a new UUID when re-creating the event. The old UUID stuck in the queue will hit this. I'll update the comment to:
```
So the event may have been removed or re-created since then.
```